src/Moq/Linq/MockSetupsBuilder.cs

```diff
@@ -151,7 +151,7 @@ private sealed class ReplaceMockObjectWithParameter : ExpressionVisitor
 
 			protected override Expression VisitMember(MemberExpression node)
 			{
-				if (node.Expression is ParameterExpression pe && pe.Type.IsDefined(typeof(CompilerGeneratedAttribute)))
+				if (node.Expression is ParameterExpression pe && pe.Type.IsDefined(typeof(CompilerGeneratedAttribute)) && pe.Type.Name.Contains("f__AnonymousType"))
```

Comment:
For anyone wondering whether we can rely on that magic string to be present for anonymous types, this is based on Roslyn source code: [`/src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs`](https://github.com/dotnet/roslyn/blob/cf55f3a58e47298426fa971d3bd9d8857c746c65/src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs#L64-L73).